array intersect

krleonid opened the pull request:
- Implement the list_intersect function with performance optimization
- Use a smaller list for the hash set, a larger one for iteration
- Add comprehensive tests

krleonid committed:
Add list_intersect function implementation

- Implement list_intersect function with performance optimization
- Use smaller list for hash set, larger for iteration

krleonid committed:
fix tidy-fix

krleonid committed:
fix tidy-fix

krleonid committed:
fix boolean & numeber intersect

krleonid committed:
fix NULL values

krleonid committed:
fix NULL values

lnkuiper reviewed (changes requested):
Thanks for the PR! I have left a comment below, and I also think the original `list_intersect` macro should be removed. You can find these in `default_functions.cpp`, which has to entries: `list_intersect` and `array_intersect`.

Maxxen commented on src/function/scalar/list/list_intersect.cpp:
```diff
@@ -0,0 +1,258 @@
+	auto original_left_child_type = ListType::GetChildType(arguments[0]->return_type);
+
+	// Set the return type to preserve the left list's element type
+	bound_function.return_type = LogicalType::LIST(original_left_child_type);
```
Type inference (in this case template resolution) will happen before any function bind callbacks are called, so Im not sure if you even need ListIntersectBind at all

krleonid committed:
fix the tests

lnkuiper reviewed (approved):
Thanks! LGTM once CI passes

taniabogatsch commented on test/sql/function/list/list_intersect.test:
```diff
@@ -139,3 +139,53 @@ SELECT list_sort(list_intersect([1, 4, 3, 5, 5, 2, 2], [5, 5, 5, 1, 1, 2, 4]));
+
+# Test type preservation - preserve left list's element type
+query I
+SELECT list_intersect([true, false], [1, 0]);
```
That being said, we can wrap all outputs in this test file in a `list_sort` to guarantee the tests don't break, but we can also just patch it, if they do (and then wonder why it broke in the first place).

taniabogatsch commented on src/function/scalar/list/list_intersect.cpp:
```diff
@@ -0,0 +1,201 @@
+	const OrderModifiers order_modifiers(OrderType::ASCENDING, OrderByNullType::NULLS_LAST);
+
+	CreateSortKeyHelpers::CreateSortKey(l_child, l_size, order_modifiers, l_sortkey_vec);
+	CreateSortKeyHelpers::CreateSortKey(r_child, r_size, order_modifiers, r_sortkey_vec);
```
Ah very smart! I was wondering how it works for nested list elements. 🤯

taniabogatsch reviewed (changes requested):
Hi @krleonid ! Sorry to last-minute chime in here. PR is looking good, I left a few minor comments. I am mostly showing up because I noticed that we originally found this issue when adding the `RealNest` benchmark.

Due to the original complexity of `list_intersect`, we could not include query 10 of that benchmark, as it would time out. Since we now have a fix, I thought it might be nice to add that benchmark back in? You can find the benchmark files in `benchmark/realnest`, and here's the file for query 10:

```sql
# name: benchmark/realnest/10_list_intersect_hashtags.benchmark
# description: list_intersect to get common hashtags
# group: [realnest]

name list_intersect_hashtags
group real_nest

require json

require httpfs

cache real_nest.duckdb

load benchmark/realnest/load.sql

run
WITH A AS (
    SELECT rowid, data.text AS text, list_transform(data.entities.hashtags, x -> x.tag) AS hashtag_list from twitter)
SELECT left_.text, right_.text, len(list_intersect(left_.hashtag_list, right_.hashtag_list)) AS common_hashtags
FROM A left_, A right_
    WHERE common_hashtags > 0 AND left_.rowid < right_.rowid 
    ORDER BY common_hashtags DESC;
```

krleonid commented on src/function/scalar/list/list_intersect.cpp:
```diff
@@ -0,0 +1,201 @@
+
+	if (l_vec.GetType().id() != LogicalTypeId::LIST || r_vec.GetType().id() != LogicalTypeId::LIST) {
+		throw InvalidInputException("list_intersect requires two LIST arguments");
+	}
```
You’re completely right - I removed this validation.

krleonid committed:
Add list_intersect hashtags benchmark and update tests

- Add benchmark for list_intersect with hashtags use case
- Remove redundant type check in list_intersect (handled by bind)
- Update tests to use list_sort for deterministic ordering

soerenwolfers commented:
If you feel extra motivated, `list_has_all` and `list_has_any` are two more currently-quadratic-but-ideally-linear macros that could be turned into proper functions in a follow-up PR.

krleonid committed:
Format benchmark file header metadata

taniabogatsch commented:
> If you feel extra motivated, list_has_all and list_has_any are two more currently-quadratic-but-ideally-linear macros that could be turned into proper functions in a follow-up PR.

@soerenwolfers - IIRC we already refactored them a while back?
